twake/backend/node/src/services/previews/services/engine/service.ts

```diff
@@ -0,0 +1,98 @@
+import { PreviewPubsubHandler, PreviewServiceAPI } from "../../api";
+import { logger, TwakeContext } from "../../../../core/platform/framework";
+import _ from "lodash";
+import { PubsubServiceAPI } from "../../../../core/platform/services/pubsub/api";
+import { PreviewPubsubCallback, PreviewPubsubRequest, ThumbnailResult } from "../../types";
+import { getTmpFile } from "../../utils";
+import fs from "fs";
+import { unlink } from "fs/promises";
+import StorageAPI from "../../../../core/platform/services/storage/provider";
+
+/**
+ * Generate thumbnails when the upload is finished
+ */
+export class PreviewProcessor
+  implements PreviewPubsubHandler<PreviewPubsubRequest, PreviewPubsubCallback> {
+  constructor(
+    service: PreviewServiceAPI,
+    private pubsub: PubsubServiceAPI,
+    readonly storage: StorageAPI,
+  ) {
+    this.service = service;
+  }
+  service: PreviewServiceAPI;
+  init?(context?: TwakeContext): Promise<this> {
+    throw new Error("Method not implemented.");
+  }
+
+  readonly topics = {
+    in: "services:preview",
+    out: "services:preview:callback",
+  };
+
+  readonly options = {
+    unique: true,
+    ack: true,
+  };
+
+  name = "PreviewProcessor";
+
+  validate(message: PreviewPubsubRequest): boolean {
+    return !!(message && message.document && message.output);
+  }
+
+  async process(message: PreviewPubsubRequest): Promise<PreviewPubsubCallback> {
+    logger.info(`${this.name} - Processing preview generation ${message.document.id}`);
+
+    if (!this.validate(message)) {
+      throw new Error("Missing required fields");
+    }
+    //Download original file
+    const readable = await this.storage.read(message.document.path, {
+      totalChunks: message.document.chunks,
+      encryptionAlgo: message.document.encryption_algo,
+      encryptionKey: message.document.encryption_key,
+    });
+    const inputPath = getTmpFile();
+    const writable = fs.createWriteStream(inputPath);
+    readable.pipe(writable);
+    await new Promise(r => {
+      writable.on("finish", r);
+    });
+    writable.end();
```

Comment:
This looks strange. Are you sure it works?

Reply:
Yes it works, maybe there is a better way to do it. For now I'm using that.

https://stackoverflow.com/questions/33599688/how-to-use-es8-async-await-with-streams